File: crates/agime-team-server/web-admin/src/utils/filePreview.ts
const EXTENSION_MIME_TYPES: Record<string, string> = {
  avif: "image/avif",
  bmp: "image/bmp",
  c: "text/x-c",
  cpp: "text/x-c++",
  css: "text/css",
  csv: "text/csv",
  doc: "application/msword",
  docm: "application/vnd.ms-word.document.macroenabled.12",
  docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  gif: "image/gif",
  go: "text/x-go",
  h: "text/x-c",
  hpp: "text/x-c++",
  htm: "text/html",
  html: "text/html",
  ico: "image/x-icon",
  java: "text/x-java",
  jpeg: "image/jpeg",
  jpg: "image/jpeg",
  js: "text/javascript",
  jsx: "text/javascript",
  json: "application/json",
  log: "text/plain",
  m4a: "audio/mp4",
  markdown: "text/markdown",
  md: "text/markdown",
  mov: "video/quicktime",
  mp3: "audio/mpeg",
  mp4: "video/mp4",
  ogg: "audio/ogg",
  pdf: "application/pdf",
  png: "image/png",
  pot: "application/vnd.ms-powerpoint",
  potx: "application/vnd.openxmlformats-officedocument.presentationml.template",
  pps: "application/vnd.ms-powerpoint",
  ppsx: "application/vnd.openxmlformats-officedocument.presentationml.slideshow",
  ppt: "application/vnd.ms-powerpoint",
  pptm: "application/vnd.ms-powerpoint.presentation.macroenabled.12",
  pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  py: "text/x-python",
  rs: "text/x-rust",
  rtf: "application/rtf",
  scss: "text/x-scss",
  svg: "image/svg+xml",
  ts: "text/x-typescript",
  tsx: "text/x-typescript",
  txt: "text/plain",
  wav: "audio/wav",
  webm: "video/webm",
  webp: "image/webp",
  xls: "application/vnd.ms-excel",
  xlsb: "application/vnd.ms-excel.sheet.binary.macroenabled.12",
  xlsm: "application/vnd.ms-excel.sheet.macroenabled.12",
  xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  xml: "application/xml",
  yaml: "application/x-yaml",
  yml: "application/x-yaml",
};

const GENERIC_BINARY_MIME_TYPES = new Set([
  "application/octet-stream",
  "binary/octet-stream",
]);

export function inferMimeTypeFromName(name: string): string {
  const clean = name.split(/[?#]/, 1)[0]?.trim().toLowerCase() || "";
  const match = /\.([a-z0-9]+)$/.exec(clean);
  if (!match) {
    return "";
  }
  return EXTENSION_MIME_TYPES[match[1]] || "";
}

export function normalizePreviewMimeType(
  mimeType: string | null | undefined,
  fileNameOrPath: string,
): string {
  const normalized = (mimeType || "").split(";", 1)[0].trim().toLowerCase();
  if (normalized && !GENERIC_BINARY_MIME_TYPES.has(normalized)) {
    return normalized;
  }
  return inferMimeTypeFromName(fileNameOrPath) || normalized;
}

export function isBrowserPreviewableFile(
  fileNameOrPath: string,
  mimeType?: string | null,
): boolean {
  const mime = normalizePreviewMimeType(mimeType, fileNameOrPath);
  if (
    mime.startsWith("text/") ||
    mime.startsWith("image/") ||
    mime.startsWith("audio/") ||
    mime.startsWith("video/")
  ) {
    return true;
  }
  return (
    mime === "application/json" ||
    mime === "application/pdf" ||
    mime === "application/msword" ||
    mime === "application/rtf" ||
    mime === "application/xml" ||
    mime === "application/x-yaml" ||
    mime === "application/vnd.ms-excel" ||
    mime === "application/vnd.ms-powerpoint" ||
    mime === "application/vnd.openxmlformats-officedocument.presentationml.presentation" ||
    mime === "application/vnd.openxmlformats-officedocument.presentationml.slideshow" ||
    mime === "application/vnd.openxmlformats-officedocument.presentationml.template" ||
    mime === "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" ||
    mime === "application/vnd.openxmlformats-officedocument.wordprocessingml.document" ||
    mime.startsWith("application/vnd.ms-excel.") ||
    mime.startsWith("application/vnd.ms-powerpoint.") ||
    mime.startsWith("application/vnd.ms-word.")
  );
}

File: crates/agime-team-server/src/agent/file_preview.rs
use mime_guess::MimeGuess;

fn is_generic_content_type(content_type: &str) -> bool {
    content_type.is_empty()
        || matches!(
            content_type,
            "application/octet-stream" | "binary/octet-stream"
        )
}

pub fn normalize_preview_content_type(path: &str, content_type: &str) -> String {
    let normalized = content_type
        .split(';')
        .next()
        .unwrap_or_default()
        .trim()
        .to_ascii_lowercase();
    if !is_generic_content_type(&normalized) {
        return normalized;
    }
    MimeGuess::from_path(path)
        .first_raw()
        .map(ToOwned::to_owned)
        .unwrap_or_else(|| {
            if normalized.is_empty() {
                "application/octet-stream".to_string()
            } else {
                normalized
            }
        })
}

pub fn workspace_preview_supported(path: &str, content_type: &str) -> bool {
    let lowered_path = path.to_ascii_lowercase();
    let content_type = normalize_preview_content_type(path, content_type);
    content_type.starts_with("text/")
        || matches!(
            content_type.as_str(),
            "application/json"
                | "application/pdf"
                | "application/msword"
                | "application/rtf"
                | "application/xml"
                | "application/x-yaml"
                | "application/vnd.ms-excel"
                | "application/vnd.ms-powerpoint"
                | "application/vnd.openxmlformats-officedocument.presentationml.presentation"
                | "application/vnd.openxmlformats-officedocument.presentationml.slideshow"
                | "application/vnd.openxmlformats-officedocument.presentationml.template"
                | "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
                | "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
                | "image/svg+xml"
        )
        || content_type.starts_with("image/")
        || content_type.starts_with("audio/")
        || content_type.starts_with("video/")
        || content_type.starts_with("application/vnd.ms-excel.")
        || content_type.starts_with("application/vnd.ms-powerpoint.")
        || content_type.starts_with("application/vnd.ms-word.")
        || lowered_path.ends_with(".csv")
        || lowered_path.ends_with(".json")
        || lowered_path.ends_with(".md")
        || lowered_path.ends_with(".markdown")
        || lowered_path.ends_with(".txt")
        || lowered_path.ends_with(".html")
        || lowered_path.ends_with(".htm")
        || lowered_path.ends_with(".svg")
        || lowered_path.ends_with(".doc")
        || lowered_path.ends_with(".docm")
        || lowered_path.ends_with(".docx")
        || lowered_path.ends_with(".rtf")
        || lowered_path.ends_with(".xls")
        || lowered_path.ends_with(".xlsm")
        || lowered_path.ends_with(".xlsx")
        || lowered_path.ends_with(".ppt")
        || lowered_path.ends_with(".pptm")
        || lowered_path.ends_with(".pptx")
        || lowered_path.ends_with(".png")
        || lowered_path.ends_with(".jpg")
        || lowered_path.ends_with(".jpeg")
        || lowered_path.ends_with(".gif")
        || lowered_path.ends_with(".webp")
        || lowered_path.ends_with(".avif")
        || lowered_path.ends_with(".bmp")
        || lowered_path.ends_with(".ico")
        || lowered_path.ends_with(".mp3")
        || lowered_path.ends_with(".m4a")
        || lowered_path.ends_with(".ogg")
        || lowered_path.ends_with(".wav")
        || lowered_path.ends_with(".mp4")
        || lowered_path.ends_with(".mov")
        || lowered_path.ends_with(".webm")
}
